diff --git a/src/cli/main.c b/src/cli/main.c
index e579a74..afc469c 100644
--- a/src/cli/main.c
+++ b/src/cli/main.c
@@ -21,6 +21,7 @@ int main(int argc, char *argv[]) {
 				"physmap",
 				"struct",
 				"physrw",
+				"IOSurface",
 				NULL,
 				NULL,
 				NULL,
@@ -85,4 +86,4 @@ sets[idx] = NULL;
 		}
 	}
 	return 0;
-}
\ No newline at end of file
+}
diff --git a/src/common.c b/src/common.c
index d78acfe..4905eef 100644
--- a/src/common.c
+++ b/src/common.c
@@ -1261,6 +1261,131 @@ static uint64_t xpf_find_iorvbar(void)
 	return iorvbar;
 }
 
+static uint64_t xpf_find_IOMemoryDescriptor_withAddressRanges_ref(void)
+{
+	__block uint64_t stringAddr = 0;
+	PFStringMetric *stringMetric = pfmetric_string_init("IOSurface: Address ranges do not cover requested allocation size\n");
+	if (gXPF.kernelIOSurfaceOsLogSection) {
+		pfmetric_run(gXPF.kernelIOSurfaceOsLogSection, stringMetric, ^(uint64_t vmaddr, bool *stop) {
+			stringAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	if (!stringAddr && gXPF.kernelIOSurfaceStringSection) {
+		pfmetric_run(gXPF.kernelIOSurfaceStringSection, stringMetric, ^(uint64_t vmaddr, bool *stop) {
+			stringAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	if (!stringAddr && gXPF.kernelPrelinkTextSection) {
+		pfmetric_run(gXPF.kernelPrelinkTextSection, stringMetric, ^(uint64_t vmaddr, bool *stop) {
+			stringAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	if (!stringAddr) {
+		pfmetric_run(gXPF.kernelStringSection, stringMetric, ^(uint64_t vmaddr, bool *stop) {
+			stringAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	pfmetric_free(stringMetric);
+	XPF_ASSERT(stringAddr);
+
+	__block PFSection *textSec = NULL;
+	PFXrefMetric *xrefMetric = pfmetric_xref_init(stringAddr, XREF_TYPE_MASK_REFERENCE);
+	__block uint64_t belowRefAddr = 0;
+	if (gXPF.kernelIOSurfaceTextSection) {
+		pfmetric_run(gXPF.kernelIOSurfaceTextSection, xrefMetric, ^(uint64_t vmaddr, bool *stop) {
+			textSec = gXPF.kernelIOSurfaceTextSection;
+			belowRefAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	if (!belowRefAddr && gXPF.kernelPLKTextSection) {
+		pfmetric_run(gXPF.kernelPLKTextSection, xrefMetric, ^(uint64_t vmaddr, bool *stop) {
+			textSec = gXPF.kernelPLKTextSection;
+			belowRefAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	if (!belowRefAddr) {
+		pfmetric_run(gXPF.kernelTextSection, xrefMetric, ^(uint64_t vmaddr, bool *stop) {
+			textSec = gXPF.kernelTextSection;
+			belowRefAddr = vmaddr;
+			*stop = true;
+		});
+	}
+	pfmetric_free(xrefMetric);
+	XPF_ASSERT(belowRefAddr);
+
+	uint32_t blAnyInst = 0, blAnyMask = 0;
+	arm64_gen_b_l(OPT_BOOL(true), OPT_UINT64_NONE, OPT_UINT64_NONE, &blAnyInst, &blAnyMask);
+	uint64_t blAddr = pfsec_find_prev_inst(textSec, belowRefAddr, 50, blAnyInst, blAnyMask);
+	XPF_ASSERT(blAddr);
+	return blAddr;
+}
+
+static PFSection *xpf_IOSurface_text_section(uint64_t address)
+{
+	if (gXPF.kernelIOSurfaceTextSection && pfsec_contains_vmaddr(gXPF.kernelIOSurfaceTextSection, address)) {
+		return gXPF.kernelIOSurfaceTextSection;
+	}
+	if (gXPF.kernelPLKTextSection && pfsec_contains_vmaddr(gXPF.kernelPLKTextSection, address)) {
+		return gXPF.kernelPLKTextSection;
+	}
+	return gXPF.kernelTextSection;
+}
+
+static uint64_t xpf_find_IOSurface_ranges(void)
+{
+	uint64_t ref = xpf_item_resolve("kernelStruct.IOSurface.IOMemoryDescriptor_withAddressRanges_ref");
+	XPF_ASSERT(ref);
+	PFSection *textSec = xpf_IOSurface_text_section(ref);
+
+	uint32_t ldrX0AnyInst = 0, ldrX0AnyMask = 0;
+	arm64_gen_ldr_imm(0, LDR_STR_TYPE_UNSIGNED, ARM64_REG_X(0), ARM64_REG_ANY, OPT_UINT64_NONE, &ldrX0AnyInst, &ldrX0AnyMask);
+	uint64_t ldrAddr = pfsec_find_prev_inst(textSec, ref, 20, ldrX0AnyInst, ldrX0AnyMask);
+	XPF_ASSERT(ldrAddr);
+
+	uint64_t ldrImm = 0;
+	arm64_dec_ldr_imm(pfsec_read32(textSec, ldrAddr), NULL, NULL, &ldrImm, NULL, NULL);
+	return ldrImm;
+}
+
+static uint64_t xpf_find_IOSurface_rangeCount(void)
+{
+	uint64_t ref = xpf_item_resolve("kernelStruct.IOSurface.IOMemoryDescriptor_withAddressRanges_ref");
+	XPF_ASSERT(ref);
+	PFSection *textSec = xpf_IOSurface_text_section(ref);
+
+	uint32_t movW1ZeroInst = 0, movW1ZeroMask = 0;
+	arm64_gen_mov_imm('z', ARM64_REG_W(1), OPT_UINT64(0), OPT_UINT64(0), &movW1ZeroInst, &movW1ZeroMask);
+	uint32_t ldrW1Inst = 0, ldrW1Mask = 0;
+	arm64_gen_ldr_imm(0, LDR_STR_TYPE_UNSIGNED, ARM64_REG_W(1), ARM64_REG_ANY, OPT_UINT64_NONE, &ldrW1Inst, &ldrW1Mask);
+
+	for (int i = 0; i < 50; i++) {
+		uint64_t curAddr = ref - (i * sizeof(uint32_t));
+		uint32_t inst = pfsec_read32(textSec, curAddr);
+		if ((inst & ldrW1Mask) == ldrW1Inst) {
+			uint64_t imm = 0;
+			arm64_dec_ldr_imm(inst, NULL, NULL, &imm, NULL, NULL);
+			return imm;
+		}
+		if ((inst & movW1ZeroMask) == movW1ZeroInst) {
+			__block uint64_t imm = 0;
+			PFXrefMetric *xrefMetric = pfmetric_xref_init(curAddr, XREF_TYPE_MASK_JUMP);
+			pfmetric_run(textSec, xrefMetric, ^(uint64_t vmaddr, bool *stop) {
+				arm64_dec_ldr_imm(pfsec_read32(textSec, vmaddr - sizeof(uint32_t)), NULL, NULL, &imm, NULL, NULL);
+				*stop = true;
+			});
+			pfmetric_free(xrefMetric);
+			return imm;
+		}
+	}
+	return 0;
+}
+
 uint64_t xpf_find_namecache(uint32_t n) {
     static uint64_t nchashtbl = 0;
     static uint64_t nchashmask = 0;
@@ -1487,5 +1612,9 @@ void xpf_common_init(void)
 	xpf_item_register("kernelStruct.thread.machine_kstackptr", xpf_find_thread_machine_kstackptr, NULL);
 	xpf_item_register("kernelStruct.thread.machine_contextData", xpf_find_thread_machine_contextData, NULL);
 
+	xpf_item_register("kernelStruct.IOSurface.IOMemoryDescriptor_withAddressRanges_ref", xpf_find_IOMemoryDescriptor_withAddressRanges_ref, NULL);
+	xpf_item_register("kernelStruct.IOSurface.ranges", xpf_find_IOSurface_ranges, NULL);
+	xpf_item_register("kernelStruct.IOSurface.rangeCount", xpf_find_IOSurface_rangeCount, NULL);
+
 	xpf_item_register("kernelSymbol.iorvbar", xpf_find_iorvbar, NULL);
 }
diff --git a/src/xpf.c b/src/xpf.c
index 11e9526..d1a6766 100644
--- a/src/xpf.c
+++ b/src/xpf.c
@@ -225,6 +225,16 @@ XPFSet gTrigonSet = {
 	}
 };
 
+XPFSet gIOSurfaceSet = {
+	.name="IOSurface",
+	.supported=xpf_supported_always,
+	.metrics={
+		"kernelStruct.IOSurface.ranges",
+		"kernelStruct.IOSurface.rangeCount",
+		NULL
+	}
+};
+
 XPFSet gNameCacheSet = {
         .name="namecache",
         .supported=xpf_supported_always,
@@ -261,6 +271,7 @@ XPFSet *gSets[] = {
 	&gDevModeSet,
 	&gArm64KcallSet,
 	&gTrigonSet,
+	&gIOSurfaceSet,
 };
 
 XPF gXPF = { 0 };
@@ -345,6 +356,9 @@ int xpf_start_with_kernel_path(const char *kernelPath)
 		gXPF.kernelAMFIDataSection = xpf_pfsec_init("com.apple.driver.AppleMobileFileIntegrity", "__DATA", "__data");
 		gXPF.kernelSandboxTextSection = xpf_pfsec_init("com.apple.security.sandbox", "__TEXT_EXEC", "__text");
 		gXPF.kernelSandboxStringSection = xpf_pfsec_init("com.apple.security.sandbox", "__TEXT", "__cstring");
+		gXPF.kernelIOSurfaceTextSection = xpf_pfsec_init("com.apple.iokit.IOSurface", "__TEXT_EXEC", "__text");
+		gXPF.kernelIOSurfaceStringSection = xpf_pfsec_init("com.apple.iokit.IOSurface", "__TEXT", "__cstring");
+		gXPF.kernelIOSurfaceOsLogSection = xpf_pfsec_init("com.apple.iokit.IOSurface", "__TEXT", "__os_log");
 		gXPF.kernelInfoPlistSection = xpf_pfsec_init("com.apple.security.AppleImage4", "__TEXT", "__info_plist");
 	}
 	else {
@@ -638,4 +652,4 @@ void xpf_stop(void)
 	}
 	
 	gXPF = (struct s_XPF){ 0 };
-}
\ No newline at end of file
+}
diff --git a/src/xpf.h b/src/xpf.h
index 7845220..800a7fb 100644
--- a/src/xpf.h
+++ b/src/xpf.h
@@ -76,6 +76,9 @@ typedef struct s_XPF {
 	PFSection *kernelAMFIDataSection;
 	PFSection *kernelSandboxTextSection;
 	PFSection *kernelSandboxStringSection;
+	PFSection *kernelIOSurfaceTextSection;
+	PFSection *kernelIOSurfaceStringSection;
+	PFSection *kernelIOSurfaceOsLogSection;
 	PFSection *kernelInfoPlistSection;
 
 	XPFItem *firstItem;

